### PR TITLE
Added support for naming spark transactions (#1893)

### LINKED ISSUE #1893: Support for spark micro framework
## Is your feature request related to a problem?
We are using Spark (https://sparkjava.com) for one of our services. #1858 introduced support for creating transactions, but the name of the created transactions can be improved.

## Describe the solution you'd like
It would be nice if the transaction could be named after the route defined in spark.

## Describe alternatives you've considered
None

### PR DESCRIPTION
**Pull request by tobiasstadler**
## What does this PR do?
Fixes #1893 

## Checklist
- [x] This is an enhancement of existing features, or a new feature in existing plugins
  - [x] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/master/CHANGELOG.asciidoc)
  - [x] I have added tests that prove my fix is effective or that my feature works
  - [x] Added an API method or config option? Document in which version this will be introduced
  - [x] I have made corresponding changes to the documentation

### CODE REVIEW

File: CHANGELOG.asciidoc [hunk 1]
```diff
@@ -51,6 +51,7 @@ our agent are not vulnerable as well, as the SMTP appender was never used, this
 for exit spans - {pull}1788[#1788]
 * Only use emulated runtime attachment as fallback, remove the `--without-emulated-attach` option - {pull}1865[#1865]
 * Instrument javax.servlet.Filter the same way as javax.servlet.FilterChain {pull}1858[#1858]
+* Added support for naming spark transactions {pull}1894[#1894]
```
Comment (felixbarny) [inline CHANGELOG.asciidoc:54]: How could we make sure people don't confuse this with Apache Spark™ suport?
Maybe like this?
```suggestion
* Added support for naming sparkjava (not Apache Spark) transactions {pull}1894[#1894]
```

Also relevant in supported-technologies.md

File: apm-agent-plugins/apm-sparkjava-plugin/src/main/java/co/elastic/apm/agent/sparkjava/RoutesAdvice.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+import spark.Route;
+import spark.routematch.RouteMatch;
+
+import static co.elastic.apm.agent.impl.transaction.AbstractSpan.PRIO_HIGH_LEVEL_FRAMEWORK;
+
+public class RoutesAdvice {
+
+    @Advice.OnMethodExit(suppress = Throwable.class, inline = false)
+    public static void onExitFind(@Advice.Return RouteMatch routeMatch) {
+        Transaction transaction = GlobalTracer.get().currentTransaction();
```
Comment (felixbarny) [inline apm-agent-plugins/apm-sparkjava-plugin/src/main/java/co/elastic/apm/agent/sparkjava/RoutesAdvice.java:34]:
```suggestion
    private static final Tracer tracer = GlobalTracer.get()

    @Advice.OnMethodExit(suppress = Throwable.class, inline = false)
    public static void onExitFind(@Advice.Return RouteMatch routeMatch) {
        Transaction transaction = tracer.currentTransaction();
```

File: apm-agent-plugins/apm-sparkjava-plugin/src/main/java/co/elastic/apm/agent/sparkjava/RoutesAdvice.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        if (transaction == null || routeMatch == null) {
+            return;
+        }
+
+        String method = routeMatch.getHttpMethod().name().toUpperCase();
+        transaction.withName(method + " " + routeMatch.getMatchUri(), PRIO_HIGH_LEVEL_FRAMEWORK, false);
```
Comment (felixbarny) [inline apm-agent-plugins/apm-sparkjava-plugin/src/main/java/co/elastic/apm/agent/sparkjava/RoutesAdvice.java:40]: avoiding allocations by using the name StringBuilder
```suggestion
        StringBuilder name = transaction.getAndOverrideName(PRIO_HIGH_LEVEL_FRAMEWORK);
        if (name != null) {
            name.append(method);
            name.append(" ")
            name.append(routeMatch.getMatchUri());
        }
```

File: apm-agent-plugins/apm-sparkjava-plugin/src/main/java/co/elastic/apm/agent/sparkjava/RoutesInstrumentation.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+import static net.bytebuddy.matcher.ElementMatchers.returns;
+import static net.bytebuddy.matcher.ElementMatchers.takesArgument;
+import static net.bytebuddy.matcher.ElementMatchers.takesArguments;
+
+public class RoutesInstrumentation extends ElasticApmInstrumentation {
+
+    @Override
+    public ElementMatcher<? super NamedElement> getTypeMatcherPreFilter() {
+        return named("spark.route.Routes");
+    }
```
Comment (felixbarny) [inline apm-agent-plugins/apm-sparkjava-plugin/src/main/java/co/elastic/apm/agent/sparkjava/RoutesInstrumentation.java:40]: Not necessary if the type matcher is just depending on the name and thus is already fast.

```suggestion
```

**Review by felixbarny — COMMENTED** [review]
Thanks! A short and sweet plugin ❤️

**Review by felixbarny — APPROVED** [review]

**Comment by felixbarny** [comment]
Could you execute `ConfigurationExporterTest` so that the `disable_instrumentation` config docs gets updated to include `sparkjava`?

**Comment by tobiasstadler** [comment]
> Could you execute `ConfigurationExporterTest` so that the `disable_instrumentation` config docs gets updated to include `sparkjava`?

Sorry, I forgot that.

**Review by SylvainJuge — APPROVED** [review]

File: CHANGELOG.asciidoc [hunk 5]
```diff
@@ -53,7 +53,6 @@ for exit spans - {pull}1788[#1788]
 * Instrument javax.servlet.Filter the same way as javax.servlet.FilterChain - {pull}1858[#1858]
 * Propagate trace context headers in HTTP calls occurring from within traced exit points, for example - when using
 Elasticsearch's REST client - {pull}1883[#1883]
-* Instrument javax.servlet.Filter the same way as javax.servlet.FilterChain {pull}1858[#1858]
```
Comment (tobiasstadler) [inline CHANGELOG.asciidoc]: Thank You!

**Comment by felixbarny** [comment]
@elasticmachine run elasticsearch-ci/docs

**Pull request merged**